jvm/src/org/jetbrains/kotlin/idea/configuration/ConfigureKotlinInProjectUtils.kt

```diff
@@ -130,26 +131,24 @@ fun isModuleConfigured(moduleSourceRootGroup: ModuleSourceRootGroup): Boolean {
  *
  * DO NOT CALL THIS ON AWT THREAD
  */
-@CalledInBackground
+@RequiresBackgroundThread
 fun getModulesWithKotlinFiles(project: Project): Collection<Module> {
     if (!isUnitTestMode() && ApplicationManager.getApplication().isDispatchThread) {
         LOG.error("getModulesWithKotlinFiles could be a heavy operation and should not be call on AWT thread")
     }
 
-    if (!project.runReadActionInSmartMode {
-            !project.isDisposed &&
-                    FileTypeIndex.containsFileOfType(KotlinFileType.INSTANCE, GlobalSearchScope.projectScope(project))
-        }) {
-        return emptyList()
-    }
-
-    return project.allModules()
-        .filter { module ->
-            project.runReadActionInSmartMode {
-                !project.isDisposed && !module.isDisposed
-                        && FileTypeIndex.containsFileOfType(KotlinFileType.INSTANCE, module.getModuleScope(true))
+    return project.runReadActionInSmartMode {
+        val projectFileIndex = ProjectFileIndex.getInstance(project)
+        val modules = mutableSetOf<Module>()
+        val ktFileProcessor = { ktFile: VirtualFile ->
+            if (projectFileIndex.isInSourceContent(ktFile)) {
+                modules.addIfNotNull(projectFileIndex.getModuleForFile(ktFile))
             }
+            true
         }
+        FileTypeIndex.processFiles(KotlinFileType.INSTANCE, ktFileProcessor, GlobalSearchScope.projectScope(project))
```

Comment:
to be honest I don't like an idea to process all kotlin files of project under readAction - is it possible to eliminate it?

Reply:
Not sure, I'll keep thinking about it.

But note, I think the read actions from before were just as long as this one because `FileTypeIndex.containsFileOfType` _also_ processes all Kotlin files in the project (even when using module scope).

But I agree things might still be improved further. E.g. maybe `KotlinConfigurationCheckerService` could be written in a way that does not require using `getModulesWithKotlinFiles` at all (not sure).